File: backend-go/internal/room/manager.go
package room

import (
	"crypto/rand"
	"encoding/hex"
	"strings"
	"sync"
)

type Manager struct {
	mu    sync.RWMutex
	rooms map[string]*Room
}

func NewManager() *Manager {
	return &Manager{
		rooms: make(map[string]*Room),
	}
}

// CreateRoom creates a new room with a random code
func (m *Manager) CreateRoom(ownerID string) *Room {
	m.mu.Lock()
	defer m.mu.Unlock()

	code := m.generateCode()
	room := NewRoom(code, ownerID)
	m.rooms[code] = room

	return room
}

// GetRoom returns a room by code (case-insensitive)
func (m *Manager) GetRoom(code string) *Room {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.rooms[strings.ToLower(code)]
}

// RemoveRoom removes a room (case-insensitive)
func (m *Manager) RemoveRoom(code string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	delete(m.rooms, strings.ToLower(code))
}

// CleanupEmptyRooms removes all empty rooms
func (m *Manager) CleanupEmptyRooms() {
	m.mu.Lock()
	defer m.mu.Unlock()

	for code, room := range m.rooms {
		if room.IsEmpty() {
			delete(m.rooms, code)
		}
	}
}

// generateCode generates a random 6-character room code
func (m *Manager) generateCode() string {
	for {
		bytes := make([]byte, 3)
		rand.Read(bytes)
		code := hex.EncodeToString(bytes)
		code = code[:6]

		// Make sure code doesn't exist
		if _, exists := m.rooms[code]; !exists {
			return code
		}
	}
}

// RoomCount returns the number of active rooms
func (m *Manager) RoomCount() int {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return len(m.rooms)
}

File: backend/src/room/manager.ts
import { Room } from './room.js';

export class RoomManager {
  private rooms: Map<string, Room> = new Map();

  // Create a new room with a random code
  createRoom(ownerId: string): Room {
    const code = this.generateCode();
    const room = new Room(code, ownerId);
    this.rooms.set(code, room);
    return room;
  }

  // Get a room by code (case-insensitive)
  getRoom(code: string): Room | undefined {
    return this.rooms.get(code.toLowerCase());
  }

  // Remove a room
  removeRoom(code: string): void {
    this.rooms.delete(code.toLowerCase());
  }

  // Cleanup empty rooms
  cleanupEmptyRooms(): void {
    for (const [code, room] of this.rooms) {
      if (room.isEmpty()) {
        this.rooms.delete(code);
      }
    }
  }

  // Generate a random 6-character room code
  private generateCode(): string {
    const chars = '0123456789abcdef';
    let code: string;
    do {
      code = '';
      for (let i = 0; i < 6; i++) {
        code += chars[Math.floor(Math.random() * chars.length)];
      }
    } while (this.rooms.has(code));
    return code;
  }

  // Get the number of active rooms
  roomCount(): number {
    return this.rooms.size;
  }
}
